src/Dependencies/PooledObjects/ArrayBuilder.cs

```diff
@@ -71,6 +71,29 @@ public ImmutableArray<T> ToImmutable()
             return _builder.ToImmutable();
         }
 
+        /// <summary>
+        /// Realizes the array and clears the collection.
+        /// </summary>
+        public ImmutableArray<T> MoveToImmutable()
+        {
+            ImmutableArray<T> result;
+            if (Count == 0)
+            {
+                result = ImmutableArray<T>.Empty;
+            }
+            else if (_builder.Capacity == Count)
+            {
+                result = _builder.MoveToImmutable();
+            }
+            else
+            {
+                result = ToImmutable();
+                Clear();
```

Comment:
this path here honestly feels like it masks a bug.  normal ImmutableArray.Builder throws in this case, indicating that the client messed something up.  I think we should do the same.  In other words, MoveToImmutable is suppsoed to be for the case where you know the count and know you've filled all of that, and thus can safely and efficiently use the underlying storage.  This isn't hte case here, so it feels like it's putting different perf expectations onto this call, even though it's called the same thing as the equiv builder operation.

Reply:
This was intentional. The current method is a "move and clear" using the most efficient operation currently available. Roslyn does not use the unintuitive semantics of `ImmutableArray<T>.Builder.MoveToImmutable` and it didn't seem advantageous to introduce complexity with those semantics.

Reply:
Please rename then (maybe ToImmutableAndClear()).  FWIW, i find the semantics of MoveToImmutable very very very intuitive.  It's exactly as it says.  It takes the underlying array and creates the ImmutableArray out of it, doing it as a simple O(1) move instead of O(n) copy.  It's exception behavior is also intuitive and appropriate IMO.  The only safe way to do this is if the count/capacity are paired.  When not, youi've done something wrong and it throws.

So, please either make the semantics the same, or rename.  I'm fine with either. I'm not ok with teh exact same name having different semantics.  Tnx :)

Reply:
I renamed this to `ToImmutableAndClear`. My original objection https://github.com/dotnet/corefx/pull/180#issuecomment-69349402 to the semantics of `MoveToImmutable` has not changed in the time since.

Reply:
Thanks very much :) #resolved